skcosmo/selection/FPS.py

```diff
@@ -0,0 +1,168 @@
+        super().__init__(matrix=matrix, mixing=mixing, tolerance=tolerance, **kwargs)
+
+        self.product = self.get_product()
+
+        if idxs is not None:
+            self.idx = idxs
+        else:
+            self.idx = [np.random.randint(self.product.shape[0])]
+        self.distances = np.min([calculate_pcov_distances(
+            self.product, idx) for idx in self.idx], axis=0)
+        self.distance_when_selected = np.nan * np.zeros(self.product.shape[0])
+
+    def select(self, n):
+        """Method for FPS select based upon a product of the input matrices
+
+        Parameters
+        ----------
+        n : number of selections to make
+
+        Returns
+        -------
+        idx: list of n selections
+        """
+        if len(self.idx) > n:
+            return self.idx[:n]
+
+        # Loop over the remaining points...
+        for i in self.progress(range(len(self.idx)-1, n-1)):
+            for j in np.where(self.distances > 0)[0]:
+                self.distances[j] = min(self.distances[j],
+                                        calculate_pcov_distances(self.product,
+                                                                 self.idx[i], [j]))
+            self.idx.append(np.argmax(self.distances))
+            self.distance_when_selected[self.idx[-1]
+                                        ] = self.distances[self.idx[-1]]
+
+            if np.abs(self.distances).max() < self.tol:
+                return self.idx
+        return self.idx
+
+    @abstractmethod
+    def get_product(self):
+        """Abstract method for computing the product (C or K) of the input matrices
+        """
+        return
+
+
+class sampleFPS(_FPS):
+    def __init__(self, matrix, mixing=1.0, tolerance=1E-12, **kwargs):
+        super().__init__(matrix=matrix, mixing=mixing, tolerance=tolerance, **kwargs)
+
+    def get_product(self):
+        """
+            Creates the PCovR modified kernel distances
+            ~K = (mixing) * X X^T +
+                 (1-mixing) * Y Y^T
+
+        """
+
+        K = np.zeros((self.A.shape[0], self.A.shape[0]))
+        if self.mixing < 1:
+            K += (1 - self.mixing) * self.Y @ self.Y.T
+        if self.mixing > 0:
+            K += (self.mixing) * self.A @ self.A.T
+
+        return K
+
+
+class featureFPS(_FPS):
+    def __init__(self, matrix, mixing=1.0, tolerance=1E-12, **kwargs):
+
+        super().__init__(matrix=matrix, mixing=mixing, tolerance=tolerance, **kwargs)
+
+    def get_product(self):
+        """
+            Creates the PCovR modified covariance
+            ~C = (mixing) * X^T X +
+                 (1-mixing) * (X^T X)^(-1/2) ~Y ~Y^T (X^T X)^(-1/2)
+
+            where ~Y is the properties obtained by linear regression.
+        """
+
+        C = np.zeros((self.A.shape[1], self.A.shape[1]), dtype=np.float64)
+
+        cov = self.A.T @ self.A
+
+        if self.mixing < 1:
+            # changing these next two lines can cause a LARGE error
```

Comment:
Could you explain why as part of the comment as well?

Reply:
If you try to make it more efficient by computing few eigenvalues of `C`, the resulting selections will not be correct